legion/engine/llri/detail/validation.hpp

```diff
@@ -0,0 +1,96 @@
+/**
+ * @file validation.hpp
+ * @copyright 2021-2021 Leon Brands. All rights served.
+ * @license: https://github.com/Legion-Engine/Legion-LLRI/blob/main/LICENSE
+ */
+
+#pragma once
+#include <cstdint>
+#include <xutility>
+
+namespace llri
+{
+    namespace detail
+    {
+        /**
+         * @brief Returns true if only a single bit is set to 1.
+        */
+        constexpr bool hasSingleBit(const uint32_t mask) noexcept
+        {
+            return (mask & (mask - 1)) == 0;
+        }
+
+        /**
+         * @brief Returns true if the container contains the value
+        */
+        template <typename C, typename T>
+        constexpr bool contains(C container, T value)
+        {
+            return std::find(container.begin(), container.end(), value) != container.end();
+        }
+    }
+}
+
+#ifdef LLRI_DISABLE_VALIDATION
+#define LLRI_DETAIL_VALIDATION_REQUIRE(param, ret)
+#define LLRI_DETAIL_VALIDATION_REQUIRE_ITER(param, i, ret)
+#define LLRI_DETAIL_VALIDATION_REQUIRE_IF(condition, param, ret)
+#define LLRI_DETAIL_VALIDATION_REQUIRE_MESSAGE(param, message)
+
+#else
+#define LLRI_DETAIL_ENABLE_VALIDATION
+
+#define LLRI_DETAIL_VALIDATION_REQUIRE(param, ret) { \
+        /* make sure the expression is executed only once */ \
+        bool requireParamResult = param; \
+        if (!requireParamResult) \
```

Comment:
how does this make sure that the expression is only executed once?

Reply:
because `param` is usually a function call so doing

`
if (!(param))
    error((param))
`

would cause the call to be executed twice

Reply:
ah yeah makes sense